zio/zjsonio/types.go

```diff
@@ -2,169 +2,159 @@ package zjsonio
 
 import (
 	"errors"
+	"fmt"
 
-	"github.com/brimdata/zed/pkg/joe"
-	"github.com/brimdata/zed/zcode"
-	"github.com/brimdata/zed/zio/tzngio"
+	"github.com/brimdata/zed/compiler/ast"
 	"github.com/brimdata/zed/zng"
 	"github.com/brimdata/zed/zson"
 )
 
-func encodeTypeAny(in zng.Type) joe.Interface {
-	if !zng.IsContainerType(in) {
-		return joe.String(tzngio.TypeString(in))
-	}
-	return encodeTypeObj(in)
-}
+type encoder map[zng.Type]string
 
-func encodeTypeObj(in zng.Type) joe.Object {
-	object := joe.NewObject()
-	typ, any := encodeType(in)
-	object["type"] = typ
-	if any != nil {
-		object["of"] = any
+func (e encoder) encodeType(zctx *zson.Context, typ zng.Type) ast.Type {
+	if name, ok := e[typ]; ok {
+		return &ast.TypeName{
+			Kind: "typename",
+			Name: name,
+		}
 	}
-	return object
-}
-
-func encodeType(typ zng.Type) (joe.String, joe.Interface) {
 	switch typ := typ.(type) {
+	case *zng.TypeAlias:
+		name := typ.Name
+		t := e.encodeType(zctx, typ.Type)
+		e[typ] = name
+		return &ast.TypeDef{
+			Kind: "typedef",
+			Name: name,
+			Type: t,
```

Comment:
```suggestion
		e[typ] = name
		return &ast.TypeDef{
			Kind: "typedef",
			Name: name,
			Type: e.encodeType(zctx, typ.Type),
```

Reply:
This might not be clear in the spec but if the type name changes the later visitor in the depth-first traversal should win so I think it's better to leave it as it is..